v2/tools/asoctl/internal/crd/cleaner.go

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package crd
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	apiextensionsV1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func CleanDeprecatedCRDVersions(ctx context.Context, cl apiextensionsV1.CustomResourceDefinitionInterface) error {
+	var crdRegexp = regexp.MustCompile(".*\\.azure\\.com")
+	var deprecatedVersionRegexp = regexp.MustCompile("v1alpha1api.{8}.*")
+
+	list, err := cl.List(ctx, v1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if list.Items == nil || len(list.Items) == 0 {
+		fmt.Println("found 0 results, make sure you have ASO CRDs installed")
+	}
+
+	var updated bool
+	for _, item := range list.Items {
+		crdName := item.Name
+
+		if !crdRegexp.MatchString(crdName) {
+			continue
+		}
+
+		crd, err := cl.Get(ctx, crdName, v1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		found, newStoredVersions := getNewVersionSet(crd.Status.StoredVersions, deprecatedVersionRegexp)
+
+		if found {
+			crd.Status.StoredVersions = newStoredVersions
+			crd, err = cl.UpdateStatus(ctx, crd, v1.UpdateOptions{})
```

Comment:
How do we know that executing this is safe? IMO we need to consider the following scenarios:
1. `asoctl crd clean` is run against a cluster that is running `alpha.6`. There are no `beta` storage versions, so no alpha storage versions should be removed. This could be protected against by a check above saying `if len(newStoredVersions) == 0 then error`.
2. Do we know that it's safe to remove the alpha stored version without forcing the version stored in etcd to be the beta version? I know that's what the cert-manager tool does and we're not doing it here, I believe because of some testing you did that suggested it wasn't needed - but it would be good to document what that testing was here and explain how we know it's OK to do that in a comment.

Reply:
do we want `len(newStoredVersions) == 0 then error` or `len(newStoredVersions) == 0 then ignore`. Later one seems fine, as there could be other CRDs which we want to update in the bulk